### ISSUE TITLE
How does purescript know which show of which instance to call ? we are not providing the name of instance anywhere

### ISSUE THREAD

**Issue by xlreon** [prasannals/learning_purescript#3]
When we are calling the show function in showShape instance we are not actually providing the name of the instance from which the show function of the class will be called.
What if we declare some other instance like

instance showInt :: Show Int where
show :: Int -> String
show x = "hello"

now when I use (show p) will it throw an error and if not how does it determine which show to call of which instance.

**Comment by prasannals** [comment]
When we say (show p), 
* The type of "p" is determined. In our case, type of "p" is Point.
* A show instance for the determined type is searched for. In our case, Show instance for Point type will be searched for.
* If found, the "show" function inside Show Point instance will be the show function which will be called.

**Comment by sriharshachilakapati** [comment]
To be more accurate, your code doesn't compile at all, and throws an error saying that it is an orphan instance. To be sure, I put out an example, so if you write this:

~~~purescript
instance showInt2 :: Show Int where
    show _ = "hello"
~~~

Then it will not compile because it is an orphan instance. Simply put, an orphan instance is a type class instance which is created in a module that neither the two types `Show` and `Int` belong to.

You can however get around this issue by wrapping it in another `data` or `newtype` so that it will be an instance of that new type instead of the Int. By the way, an instance for `Show Int` is already present in `Data.Show` module which is re-exported by `Prelude` module.

**Comment by xlreon** [comment]
Okay, thanks for the explanation now I am able to understand it better.

**Comment by sriharshachilakapati** [comment]
In response to your second question, ie., why do instances in PureScript have names, I have got this question for myself a while ago, and I went onto StackOverflow for it. Here is the question I've asked:

[Why do instances in PureScript have names?](https://stackoverflow.com/questions/48151439/why-do-instances-in-purescript-have-names)

The answer I got says that it is because of readability of generated JavaScript code. To see it with more details, let's assume that we have the following example:

~~~purescript
class ToString a where
    toString :: a -> String
~~~

And now we have a type for point and a ToString instance of it.

~~~purescript
data Point2D = Point2D Int Int

instance point2DToString :: ToString Point2D where
    toString (Point2D x y) = "(" <> (show x) <> ", " <> (show y) <> ")"
~~~

Now we write this one line to log a point.

~~~purescript
log $ toString (Point2D 3 2)
~~~

It will generate this JS code.

~~~javascript
Control_Monad_Eff_Console.log(toString(point2DToString)(new Point2D(3, 2)));
~~~

Did you get it now? Normally you would have thought that you are passing only the point to the `toString` function, but actually you see that you are passing your type class instance to it. This is why the PureScript compiler requires you to name the instances unlike Haskell.

So wonder how the `toString` function looks like?

~~~javascript
var ToString = function (toString) {
    this.toString = toString;
};
var toString = function (dict) {
    return dict.toString;
};
~~~

Now where is the instance? It's here.

~~~javascript
var point2DToString = new ToString(function (v) {
    return "(" + (Data_Show.show(Data_Show.showInt)(v.value0) + (", " + (Data_Show.show(Data_Show.showInt)(v.value1) + ")")));
});
~~~

That is how it is like. So passing it to your `toString` function will make it use the instance function to call the value with. I know it's a bit complex like this, but this is exactly how it functions. Hope it is more clear now.
